### ISSUE TITLE
Changes to public properties cause an exception

### ISSUE THREAD

**Issue by code-distortion** [livewire/livewire#2718]
I'd like to start by saying that I appreciate your work Caleb and the community that's grown around Livewire.

This issue is about problem I've found that I'd like to get further input about.

Currently (as of ver 2.4.0 at least), when visiting a page with a Livewire component in it; If a public class property is then added to the component, the user will be shown an ErrorException if they continue to interact with it. The same occurs when removing a public property,  or changing their order.

I proposed a [PR aiming to make Livewire tolerant](https://github.com/livewire/livewire/pull/2584) of these changes.  It updated the parts of Livewire that deal with reading the data as it comes in from an ajax request,  and the javascript that reads the response from the request.

The PR description contains further details and example code to illustrate the problem.

Caleb suggested there might be a different approach that could be investigated,  so I'd like to try and continue the conversation.  I'm happy to help work on a solution.

We've been evaluating Livewire and would love to use it.  I believe this is something that will cause problems for our customers at some stage,  where users would intermittently see errors as we deploy changes that we don't have a way of handling gracefully.

Is this something that anybody has seen or dealt with already?

*If* Livewire were tolerant of changes to the properties,  developers could decide whether their changes need to be dealt with,  by:
- adding their own "$version" public property,  and an equivalent protected or private property, giving them the same starting value,
- changing the value of both when making a breaking change,
- when actions are triggered,  a check could be made that compares the value of the public property to the private one,
- when different,  the dev could choose how to handle the request (eg. start the process again,  display a message,  reload the page, etc)

Something like this is what I was hoping to make possible with the PR.

Following from that, it would be possible for Livewire to automate part of this process by reading a $version property dedicated for this purpose, and perhaps call a particular method when it detects a change.  The value could be added as part of the request meta-data instead being bundled with the user-land data.

Regarding *automatic* detection of changes.  Here are some thoughts:
- Changes to class properties *might* represent a "version" change of that component.  They might not as well,
- Changes to logic that the class carries out *might* also represent a "version" change, but this would be hard (or impossible) to detect,
- It might be possible to catch these exceptions at Laravel's exception-handling level,  and deal with it in a more cover-all way. But exceptions more specific than the current ErrorException would need to be triggered so they could be detected.

Thanks for reading all of this.

Does anybody have any ideas or suggestions?

@calebporzio: Is this something you'd consider addressing in Livewire v3?

**Comment by austenc** [comment]
@code-distortion Thanks for bringing this up again and being willing to work on it. It has been brought up before, but this is the only issue I could find while digging around. I think there might have been one or two other issues, too.

https://github.com/livewire/livewire/issues/2679

As far as a solution, maybe it should be automatic so that developers don't even need to be concerned about it? I do like the idea of Livewire providing an event and/or callback function to allow customization when a component is updated.

What if instead of tracking individual properties, Livewire tracked some sort of "fingerprint" of the basic state of a component, and just forced a reload when that fingerprint changed?

**Comment by calebporzio** [comment]
Thanks @code-distortion, I can see supporting this in the current version after we think it through a bit more.

Just for the same of exploration, of we DID go a different route where livewire detects a new deployment by a unique key or something, my primary question is:

What is the user experience when a new deployment happens? Are they prompted and forced to restart the page?

That doesn't seem great, wondering if there are better ideas?

Also, are there other scenarios where currently a new deployment would cause errors BESIDES the public property thing addresses in this PR?

**Comment by code-distortion** [comment]
Hi Austen and Caleb.

Yes,  if the property errors were solved,  I can see other errors still being possible.  For example,  code inside a component might generate an error if it made assumptions about property values that aren't true in requests containing pre-deployment data.

---

If Livewire could detect a breaking change *somehow*,  I'm not sure there'd be a "best" action to take.  I imagine it'd be a development preference.

The user should be stopped (as in,  it's a "breaking" change so they shouldn't be able to proceed).

The process could be looked at by breaking it down in to parts.  I've tried to add a few extra options:

User interaction:
- display a *custom* message,  and let the user click a link or button to trigger an outcome below,
- display a *generic* message,  and let the user click a link or button to trigger an outcome below,
- simply do the outcome straight away.

Outcomes:
- reset the state of the component,  to essentially start the process again,
- reload the page,
- emit an event,
- forward the user somewhere else perhaps.

Livewire might be able to help automate the *outcomes*,  and could trigger them straight away if desired.  It might be able to help *aid* the process of showing a message if that's desired.  I imagine it'd be important to let the developer control what it looks like.

If we were to look at doing these,  the variations would start to make this a fiddly problem to solve.

An alternative might be to tell the dev that a change was detected (via some "*a-change-has-happened*" method),  and let them handle the process.  With the exception of the "reset component" option,  the outcomes mentioned above are available in Livewire already.

---

Regarding *how* to detect a breaking change;  I think it would be possible to take a sort of fingerprint of a Livewire class itself,  perhaps using reflection,  but think it would be a blunt instrument.

Changes to the fingerprint would sometimes represent a breaking change,  but it'd be hard to know for sure.  I wouldn't want to make users restart a process when it wasn't really needed.

And there are things that are hard or impossible to detect.  For example,  in our application,  we would want Livewire to call business-logic code that's kept outside of the component.  Changes to component blade templates might also be breaking changes.

My feeling is that it would be more worthwhile to let the developer choose when a change has occurred with some sort of version identifier.

---

Jumping back for a moment.  The https://github.com/livewire/livewire/issues/2679 issue and PR you mentioned Austen,  interestingly looked to fix one of the same things I was looking at - when a new property is added to a Livewire class.

@calebporzio - Would it be more manageable to create separate PR's,  one at a time,  for each of the 3 situations (new property, removed property,  change in property order)?  I could produce tests for each and we wouldn't rush the process.
